### ISSUE TITLE
FR: Carthage support

### ISSUE THREAD

**Issue by okydk** [firebase/firebase-ios-sdk#9]
Now that the Firebase SDKs are open source, can we then have Carthage support? 😍 
```
*** Cloning firebase-ios-sdk
*** Checking out firebase-ios-sdk at "4.0.0"
*** Skipped building firebase-ios-sdk due to the error:
Dependency "firebase-ios-sdk" has no shared framework schemes
```

**Comment by morganchen12** [comment]
We can't add Carthage support via a shared framework scheme for the time being since all the Firebase libraries have a common dependency in Analytics, which is still closed source and vended as a static library. If we were to try to distribute the other Firebase libs as dynamic frameworks, they would all have link time dependencies on Analytics and would end up duplicating the Analytics classes in the runtime, bloating app size and causing undefined behavior in some instances.

Carthage does support vending binaries, so we could still distribute these as static binaries with static dependencies on Analytics. Others wouldn't be able to easily build and redistribute libraries wrapping Firebase, but otherwise we'd retain many of the benefits of using Carthage. But before this can happen Analytics also needs to be distributed through Carthage, since mixing dependency managers is not something we should encourage developers to do.

The most straightforward solution would be to distribute Analytics as a dynamic framework through Carthage and then build the open source Firebase frameworks on top of that via a shared framework scheme. This is unlikely to happen in the very short term. In the longer term, we'd need to weigh the drawback of causing longer app launch times for dylib users versus the benefit of allowing users to install Firebase through Carthage and easily build frameworks wrapping Firebase. If you have either of these concerns (app launch times/want to build on top of Firebase), please share your thoughts!

**Comment by okydk** [comment]
Ok, fingers crossed for support in the future 🤞  and thanks for the clarification.

**Comment by soheilbm** [comment]
For time being I already created a carthage compatible for firebase. https://github.com/soheilbm/firebase

**Comment by mhuusko5** [comment]
@morganchen12 I'm very much interested in building frameworks on top of Firebase, and I think the more Firebase continues to grow as a defacto database and auth/etc. solution, the more people/companies are going to be wrapping it, at least for internal use (but probably convenience frameworks at some point).
At the moment though I'm most interested in this Analytics topic. From what I can tell at first pass of the source (at least of FirebaseCore), Analytics being present is desired but not required. Is this actually/will this continue to be the case?

**Comment by salqadri** [comment]
Hi @mhuusko5. Google Analytics for Firebase is what enables us to offer an integrated experience across the various other Firebase services we offer. As such we try to make Firebase Analytics included in the default experience since this is will maximize the value to our users. A great example of this is the AdMob integration with Firebase. [When used together](https://analytics.googleblog.com/2017/05/firebase-analytics-gets-new-features.html), we are able to combine metrics from both AdMob and Analytics to answer more sophisticated questions like "On which screens are users being exposed to advertising the most or the least?" However, we know that some advanced users may explicitly only want AdMob SDK which is why we avoid having hard dependencies. So yes, since this is by design, and so you can expect that we will continue to only be using weak dependencies for the majority of our inter-product integrations. There are other weak dependencies in the codebase as well, such as between Database and Auth, such that Database works without it, but when Auth is included in an app its functionality is enhanced. Hope that clarifies this.

**Comment by mhuusko5** [comment]
@salqadri fantastic, thanks for the breakdown, and I love the design strategy. 👍

**Comment by morganchen12** [comment]
@mhuusko5 for the time being you can create frameworks wrapping Firebase for your own personal use by just embedding them in a CocoaPods target without any issues. Life becomes much harder once you try to distribute the framework afterwards.

**Comment by AndrewSB** [comment]
Distributing Analytics through Carthage shouldn't be that hard. For an example, this is what mapbox does: https://www.mapbox.com/ios-sdk/

Relevant line in a consumer's cartfile would be `binary "https://www.mapbox.com/ios-sdk/Mapbox-iOS-SDK.json" ~> 3.5`, which just points to a JSON file (https://www.mapbox.com/ios-sdk/Mapbox-iOS-SDK.json ) with version numbers and a URL from where to download the .framework

**Comment by morganchen12** [comment]
@AndrewSB for reference the issue here isn't Carthage, it's that we currently don't ship Analytics as a dynamic framework, which we must before distributing it through Carthage.

**Comment by AndrewSB** [comment]
Ahh, I see. Still a static framework, I take back what I said then. So I guess we're waiting on that dynamic framework then

**Comment by paulb777** [comment]
Carthage has recently added some info to its README about integrating static frameworks that could be helpful - https://github.com/Carthage/Carthage#build-static-frameworks-to-speed-up-your-apps-launch-times

**Comment by TofPlay** [comment]
I create some dynamic frameworks compatible with carthage for my own projects with dependencies. It's not really complex. And if you want to create a cross-platform framework compatible with carthage you can take a look to articles like this one:
[Creating Cross-Platform Swift Frameworks for iOS, watchOS, and tvOS via Carthage and CocoaPods](https://goo.gl/mzZ3Si) 👍

**Comment by TofPlay** [comment]
If it's can help I create this tutorial: [Swift Cross Platform Framework](https://github.com/TofPlay/SwiftCrossPlatformFramework)

**Comment by prsolucoes** [comment]
+1

**Comment by alokc83** [comment]
+1

**Comment by Davtyanag** [comment]
+1

**Comment by terflogag** [comment]
+1

**Comment by CareSourceMichael** [comment]
+1

**Comment by iSevenDays** [comment]
Is there any progress?

**Comment by paulb777** [comment]
This PR will be updated when there is progress.

Please don't add additional comments unless you are interested in contributing to the implementation.

Use :+1: to indicate interest.

**Comment by mcgaryes** [comment]
👍🏻

**Comment by joselinoneto** [comment]
+1

**Comment by basememara** [comment]
![22u1bz](https://user-images.githubusercontent.com/892152/35042523-0e0cab9e-fb57-11e7-8370-9c8707575133.jpg)

Not all of us are ok with CocoaPods molesting our project files. Stripe, Intercom, and Facebook all have beast SDK's and have Carthage support, and you're the ones supposed to be mobile dev friendly.

The manual setup is ridiculously insane and a huge barrier. The impression I get is you got something super legacy I wouldn't touch with a 10-foot pole.

**Comment by mbuchetics** [comment]
How is this still an open issue? Is somebody actively working on it?
Carthage 0.28 supports static binaries: https://github.com/Carthage/Carthage/releases/tag/0.28.0

**Comment by localredhead** [comment]
+1

**Comment by paulb777** [comment]
@AndrewSB Thanks for digging in and glad to hear you're making progress, but I'm not sure it's currently possible with Carthage to support Firebase as the mix of source static frameworks currently required to work with the Analytics and InstanceID binary static framework dependencies.

I'm currently working on making a pure binary Firebase Carthage distribution derived from the [Firebase zip distribution](https://firebase.google.com/docs/ios/setup#frameworks) and planning to make it publicly available in the near future.

**Comment by AndrewSB** [comment]
@paulb777 thats great to hear, that does seem like the complete solution. Keep us updated on how that's going. Like I mentioned last year, it should be as simple as creating a JSON file like https://www.mapbox.com/ios-sdk/Mapbox-iOS-SDK.json once the releases are uploaded in a standard way

**Comment by schlingding** [comment]
Thanks @paulb777 !

For the time being this is my solution:
https://github.com/MonsantoCo/FirebaseFramework

**Comment by paulb777** [comment]
An experimental binary Carthage Firebase distribution is now available including all of Firebase except Firestore and Invites. Details at https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md.

I've also created a new github [Carthage label](https://github.com/firebase/firebase-ios-sdk/issues?utf8=%E2%9C%93&q=label%3ACarthage+) for any specific issues or feedback.

**Comment by olejnjak** [comment]
I think this is just fine but can we expect Firestore support in near future?

**Comment by paulb777** [comment]
@olejnjak Firestore has a dependency on gRPC which has resources and I couldn't figure out how to get the Carthage installer to install resources.  Any ideas?

**Comment by AndrewSB** [comment]
@paulb777 are you talking about https://github.com/grpc/grpc? Can you be more specific with the resources that are required? Also, the folks over at Carthage are usually pretty helpful. If you aren't sure how to do this, creating an Issue there will probably get you an answer

**Comment by paulb777** [comment]
@AndrewSB Specifically gRPC/etc/roots.pem from https://github.com/grpc/grpc. 

I couldn't find a place to put it in the zip without being ignored by the Carthage installer.

The next time I have bandwidth for Carthage, I'll take another stab at it and make an issue. In the meantime, suggestions and contributions are welcome.

**Comment by paulb777** [comment]
- I updated the Carthage distribution with Firebase 4.11.0.
- It now [includes Functions, Firestore, and Invites](https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md#carthage-usage).
- Instructions for adding the Resources in Firestore and Invites are [now available.](https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md#carthage-usage)

Thanks to @blender for [help with Carthage binary resource installation](https://github.com/Carthage/Carthage/issues/2401).

**Comment by rerelease** [comment]
We're seeing a linker warning when using Firebase with Carthage

```
ld: warning: Auto-Linking framework linker option at .../Carthage/Build/iOS/Firebase.framework/Firebase is not a dylib and not an archive
```

Any clue what's causing this? It still appears when `Link frameworks automatically` is disabled.

**Comment by olejnjak** [comment]
@rerelease do you link empty `Firebase.framework`? You shouldn't https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md#carthage-usage 
> Delete Firebase.framework from the Link Binary With Libraries Build Phase.

**Comment by Drusy** [comment]
Having the same problem @rerelease, having the `import Firebase` statement produces this warning.
Replacing with any other library works `import FirebaseCore` for example or `import FirebaseAnalytics`.

**Comment by prsolucoes** [comment]
How to add this repository to Carthage? Only need paste that lines "binary ..." to the Cartfile?

**Comment by floriangbh** [comment]
> How to add this repository to Carthage? Only need paste that lines "binary ..." to the Cartfile?

README : https://github.com/firebase/firebase-ios-sdk/blob/master/Carthage.md#carthage-usage

**Comment by paulb777** [comment]
@prsolucoes Does #1061 answer your question?

**Comment by prsolucoes** [comment]
Yes. In my opinion, i think that something more clear as "paste de following lines into your Cartfile" will be easier to carthage newbies like me auhahuahuahuahua. But i understand that is paste lines and im testing now.

**Comment by prsolucoes** [comment]
Hi,

After include this reps:

```
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json"
```

And carthage build, i have include the frameworks, but got error:

```
Invalid architecture: Could not read architectures from /Users/paulo/Developer/workspaces/xcode/ubook-reader-ios/Carthage/Build/iOS/FirebaseCoreDiagnostics.framework
Command /bin/sh failed with exit code 1
```

**Comment by kasyanov-ms** [comment]
I've just deleted that framework, and all works well

**Comment by prsolucoes** [comment]
I can ignore it? Remove it from my build and copy-framework?

**Comment by prsolucoes** [comment]
Today im using all of this:

![image](https://user-images.githubusercontent.com/395096/38575849-c83b06ac-3cd2-11e8-87cc-b90d3334b319.png)

To use Analytics and Messaging parts.

**Comment by paulb777** [comment]
@kasyanov-ms @prsolucoes I deployed Firebase 4.12.0 to Carthage and fixed the FirebaseCoreDiagnostics error in the process.

I'm not able to reproduce the "This app could not be installed at this time." issue. If that's still occurring for you, please create another issue with details on how to reproduce.

**Comment by prsolucoes** [comment]
Hi @paulb777, Thanks for the update. But FirebaseCoreDiagnostics is needly or not?

**Comment by prsolucoes** [comment]
Now with the new version everything is working and compiling, but when i run i get the error:

![image](https://user-images.githubusercontent.com/395096/38687450-1aebaf98-3e4d-11e8-9d59-7d77eea3b515.png)

**Comment by olejnjak** [comment]
@prsolucoes i think that this might be some problem with your project configuration...

**Comment by prsolucoes** [comment]
@olejnjak @paulb777  No guys, dont be a problem here. I create an empty and simple new project, add carthage file, download the frameworks, link into the project and the same problem happen. If i remove all from firebase, the project back to work:

The repository that i create only to test firebase and carthage:
https://github.com/prsolucoes/firebase

**Comment by prsolucoes** [comment]
@olejnjak @paulb777 Hi. I made more command line tool in my sample project (https://github.com/prsolucoes/firebase). After run on simulator (make run-simulator) i got error:

```
ios-sim launch ./build/Build/Products/Debug-iphonesimulator/FirebaseTest.app --devicetypeid=iPhone-7
An error was encountered processing the command (domain=IXUserPresentableErrorDomain, code=1):
This app could not be installed at this time.
Underlying error (domain=MIInstallerErrorDomain, code=57):
	Found bundle at /Users/paulo/Library/Developer/CoreSimulator/Devices/CE49D8E9-CC06-4398-9ED1-D11DB21B64C9/data/Library/Caches/com.apple.mobile.installd.staging/temp.NvZWu9/extracted/FirebaseTest.app/Frameworks/FirebaseCore.framework with the same identifier (com.firebase.Firebase) as bundle at /Users/paulo/Library/Developer/CoreSimulator/Devices/CE49D8E9-CC06-4398-9ED1-D11DB21B64C9/data/Library/Caches/com.apple.mobile.installd.staging/temp.NvZWu9/extracted/FirebaseTest.app/Frameworks/Protobuf.framework
```

I think that it can be a clue.

Can you check?

**Comment by prsolucoes** [comment]
Hi @olejnjak and @paulb777, 

I solved it changing manually all info.plist as i sad and it is working now:

```
./Carthage/Build/iOS/Protobuf.framework/Info.plist:	<string>com.firebase.Firebase.Protobuf</string>
./Carthage/Build/iOS/Firebase.framework/Info.plist:	<string>com.firebase.Firebase</string>
./Carthage/Build/iOS/FirebaseCore.framework/Info.plist:	<string>com.firebase.Firebase.Core</string>
./Carthage/Build/iOS/FirebaseAnalytics.framework/Info.plist:	<string>com.firebase.Firebase.Analytics</string>
./Carthage/Build/iOS/FirebaseMessaging.framework/Info.plist:	<string>com.firebase.Firebase.Messaging</string>
./Carthage/Build/iOS/GoogleToolboxForMac.framework/Info.plist:	<string>com.firebase.Firebase.GTFM</string>
./Carthage/Build/iOS/nanopb.framework/Info.plist:	<string>com.firebase.Firebase.nanopb</string>
./Carthage/Build/iOS/FirebaseNanoPB.framework/Info.plist:	<string>com.firebase.FirebaseNanoPB</string>
./Carthage/Build/iOS/FirebaseInstanceID.framework/Info.plist:	<string>com.firebase.Firebase.InstanceID</string>
./Carthage/Build/iOS/FirebaseCoreDiagnostics.framework/Info.plist:	<string>com.firebase.Firebase.CoreDiagnostics</string>
```

Now you only need correct all the frameworks to have individual bundle-id. What i do is change to **"com.firebase.Firebase + . + [Framework Name]"**

**Comment by rerelease** [comment]
@prsolucoes these are static frameworks, you are not supposed to add them to the `copy-frameworks` script

**Comment by prsolucoes** [comment]
Hi,

So, i dont need add anyone too copy-framework?

![image](https://user-images.githubusercontent.com/395096/38703298-45f7b714-3e79-11e8-90f8-07677dfddc6c.png)

Thanks.

**Comment by paulb777** [comment]
Right - the Firebase frameworks are all static and linked at build time and don't need to be copied for runtime linking.

**Comment by prsolucoes** [comment]
Hi,

I have updated everything and all is working.

I change all the project and make a sample that receive a message from Firebase console and show a push message. It is nice. With carthage support and makefile to help users.

Thanks for all support guys. Here is the repository:

https://github.com/prsolucoes/firebase

**Comment by mortenbekditlevsen** [comment]
Hi there,
I have a question about the Carthage support help page.

Under Static Frameworks it is stated that:
"Note that the Firebase frameworks in the distribution include static libraries. While it is fine to link these into apps, it will generally not work to depend on them from wrapper dynamic frameworks."

Could this comment be elaborated? 
Is it a generic logical issue with static libs inside dynamic frameworks?
Is it an issue with how Carthage does this?
Or is it an issue with the libraries themselves?

We are already manually building a dynamic wrapper framework using CocoaPods - and use this dynamic framework in our own Carthage setup.

Initial tests show that using this Carthage repo instead (still linking into a dynamic framework that contains all of our networking/sync logic) appears to work just fine, so it would be great to know what the issue might be.

Carthage support is a great update, thanks! Looking forward to the tvOS and macOS builds to become available through this channel as well. ;-)

**Comment by paulb777** [comment]
@mortenbekditlevsen  Thanks for the feedback!

Yes - it's a general issue about how static and dynamic linking works.  Static linking happens at build time. Dynamic linking happens at run time. if a dynamic library depends upon a static library, that static library will be loaded at runtime, which it isn't necessarily built to do.  When a second dynamic library is linked that depends on the same static library, two copies of the static library will get linked into the app.

While app's like this may pass simple tests, they will have subtle failures based on indeterministic behavior of data inside the static library - like values of static/global variables or calls to `isKindOfClass:`.

Xcode will log in the console `One of the two will be used. Which one is undefined.`

While it may be possible to make workarounds viable for particular use cases, none of Xcode, CocoaPods, or Carthage have a general purpose solution for depending on static libraries from dynamic frameworks.

**Comment by mortenbekditlevsen** [comment]
Thanks, @paulb777 ,

So am I correct in assuming, that if only one dynamic library has the static libraries linked, then you will only ever have one copy of the static libraries loaded, and you will get no inderterminism in that regard?

**Comment by morganchen12** [comment]
@mortenbekditlevsen that's correct.

**Comment by olejnjak** [comment]
I was also wondering, since Crashlytics are now part of Firebase and are the default crash reporting tool, can we expect official Carthage support? I mean different than "some guys" repo (https://github.com/Carthage/Carthage/issues/2036)?

**Comment by schlingding** [comment]
+1 for Fabric/Crashlytics inclusion. I am waiting for this inclusion before I switch from my current Firebase solution.

**Comment by TKBurner** [comment]
Todd from Crashlytics here. Right now we don't have plans for official Carthage support, but all previous workarounds should continue to work in the new system. Thanks! 👍

**Comment by konkontos** [comment]
I was told to investigate using Firebase for one an upcoming app. 
This thread has been _enlightening_ in helping me make a decision.
It seems Carthage support is officially out of the question - i.e., I'll look for another solution.

Unfortunately, I'm one of those developers not comfortable using a package manager that alters my project files / settings.

**Comment by blender** [comment]
Hello,

This [PR](https://github.com/Carthage/Carthage/pull/2436) which just landed in Carthage/master makes it possible to use the archive at https://kit-downloads.fabric.io/cocoapods/fabric/1.7.6/fabric.zip with Carthage.

What is left to do, is for Google/Crashlytics to host the binary dependency file somewhere. This is a simple JSON, see https://github.com/Carthage/Carthage/blob/master/Documentation/Artifacts.md#binary-project-specification

@TKBurner Do you think this is something that Fabic users can expect? It would definitely be better for developers to get Fabric from the original source rather than so far benevolent third parties.

**Comment by m-toyoda** [comment]
+1

**Comment by eskerber** [comment]
@TKBurner Does you mean that we can't expect Crashlytics.framework and Fabric.framework to support Carthage, but _FirebaseXXXX.framework_ and all Firebase originals _will_ have Carthage support?

**Comment by miguelrc86** [comment]
Hi everyone!

I've being trying to make `FirebaseDatabase` work with no success... I followed the installation guide throughfully, adding `ObjC` to other linker flags, removing the Firebase.framework from the linked frameworks, including `GoogleService-Info.plist` in my build target and my subset for the Cartfile is as follows: 

```
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebaseDatabaseBinary.json"
```

I'm receiving a lot of `Apple Mach-O Linker errors`... 😕 This is a new project using Xcode 9.3, iOS 11 support, Swift 4.1 any clues? Thanks in advance!

![screen shot 2018-05-09 at 22 56 36](https://user-images.githubusercontent.com/8277680/39851416-43d6b912-53dc-11e8-9002-933a9cabd324.png)

**Comment by paulb777** [comment]
@miguelrc86 Occasionally, Xcode's automatic linking fails to pull in all of the necessary libraries. You can find the missing libraries by internet searching the missing symbol errors or systematically adding in all of the libraries that show up in `Carthage/Build/iOS/*/Modules/module.modulemap`.

In this case, the c++ and sqlite libraries need to be added to the `Link Binary With Libraries` Build Phase:

![screen shot 2018-05-09 at 9 14 51 pm](https://user-images.githubusercontent.com/73870/39851823-0cad2c22-53ce-11e8-8a34-c098a5744ec2.png)

**Comment by miguelrc86** [comment]
Oh! Thank you so much @paulb777 for your ultra quick and clear response, will do that! 🙂

**Comment by pisarm** [comment]
This also helped me - I was missing `libicucore.tbd` and `libc++.tbd`.

**Comment by miguelrc86** [comment]
Thanks @paulb777 @pisarm added those libraries, also had to include StoreKit and voila! Firebase is up and running now 👊🏻

**Comment by TKBurner** [comment]
You are all in great hands with @paulb777, we had a chance to meet and are trying to see what we can as far as Crashlytics.

**Comment by jemartti** [comment]
I've got Carthage working and running—however, I'm seeing the same linker warning as @rerelease and @Drusy:

```
ld: warning: Auto-Linking framework linker option at /Users/jacob/workspace/fundu-ios/Carthage/Build/iOS/Firebase.framework/Firebase is not a dylib and not an archive
```

There was a mention of removing any `import Firebase` statements; however, that seems to contradict the required step of configuring a `FirebaseApp` shared instance (e.g. `FirebaseApp.configure()` in `application:didFinishLaunchingWithOptions:`).

Has anyone managed to get rid of this warning?

**Comment by miguelrc86** [comment]
@jemartti I resolved this by using `import FirebaseCore` instead if I remember well, so yeap, you need to follow the instructions by removing the Firebase framework and its related imports, hope it helps.

**Comment by nidegen** [comment]
I often get an error when runninc carthage update with the framework protobuf. No idea where it comes from. Happens on different machines. Any ideas? thanks:)
```
carthage update --platform iOS
*** Downloading binary-only framework FirebaseDynamicLinksBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseDynamicLinksBinary.json"
*** Downloading binary-only framework FirebasePerformanceBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json"
*** Downloading binary-only framework FirebaseABTestingBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseABTestingBinary.json"
*** Downloading binary-only framework FirebaseFunctionsBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseFunctionsBinary.json"
*** Downloading binary-only framework Fabric at "https://raw.githubusercontent.com/Building42/Specs/master/Carthage/Fabric.json"
*** Downloading binary-only framework FirebaseDatabaseBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseDatabaseBinary.json"
*** Downloading binary-only framework FirebaseMessagingBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json"
*** Downloading binary-only framework FirebaseAnalyticsBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
*** Fetching Localize-Swift
*** Downloading binary-only framework FirebaseFirestoreBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseFirestoreBinary.json"
*** Downloading binary-only framework FirebaseAdMobBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAdMobBinary.json"
*** Downloading binary-only framework FirebaseAuthBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAuthBinary.json"
*** Downloading binary-only framework Crashlytics at "https://raw.githubusercontent.com/Building42/Specs/master/Carthage/Crashlytics.json"
*** Downloading binary-only framework FirebaseRemoteConfigBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseRemoteConfigBinary.json"
*** Downloading binary-only framework FirebaseStorageBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseStorageBinary.json"
*** Downloading binary-only framework FirebaseInvitesBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseInvitesBinary.json"
*** Checking out Localize-Swift at "2.0.0"
*** Downloading binary-only framework Crashlytics at "https://raw.githubusercontent.com/Building42/Specs/master/Carthage/Crashlytics.json"
*** Downloading binary-only framework Fabric at "https://raw.githubusercontent.com/Building42/Specs/master/Carthage/Fabric.json"
*** xcodebuild output can be found in /var/folders/xw/40bn71nx71591_rwtsp05k780000gn/T/carthage-xcodebuild.3BRGy2.log
*** Downloading binary-only framework FirebaseABTestingBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseABTestingBinary.json"
*** Downloading binary-only framework FirebaseAdMobBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAdMobBinary.json"
*** Downloading binary-only framework FirebaseAnalyticsBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAnalyticsBinary.json"
*** Downloading binary-only framework FirebaseAuthBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseAuthBinary.json"
*** Downloading binary-only framework FirebaseDatabaseBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseDatabaseBinary.json"
*** Downloading binary-only framework FirebaseDynamicLinksBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseDynamicLinksBinary.json"
*** Downloading binary-only framework FirebaseFirestoreBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseFirestoreBinary.json"
*** Downloading binary-only framework FirebaseFunctionsBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseFunctionsBinary.json"
*** Downloading binary-only framework FirebaseInvitesBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseInvitesBinary.json"
*** Downloading binary-only framework FirebaseMessagingBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseMessagingBinary.json"
*** Downloading binary-only framework FirebasePerformanceBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json"
*** Downloading binary-only framework FirebaseRemoteConfigBinary at "https://dl.google.com/dl/firebase/ios/carthage/FirebaseRemoteConfigBinary.json"
A shell task (/usr/bin/xcrun dwarfdump --uuid /Users/<...>/Carthage/Build/iOS/Protobuf.framework/Protobuf) failed with exit code 1:
error: unable to open '/Users/<...>/Carthage/Build/iOS/Protobuf.framework/Protobuf': No such file or directory
```

**Comment by paulb777** [comment]
@nidegen I've seen this a few times, but not able to reproduce reliably. I suspect a Carthage race condition since `Protobuf.framework` is a dependency of multiple components.

Rerunning the `carthage update` command is a workaround.

**Comment by nidegen** [comment]
@paulb777 yes I have also just reupdated carthage a couple of times, but it is really a pain for our CI pipelines😬😬

are the binaries not just download only? how comes carthage has to check out dependencies there?

**Comment by paulb777** [comment]
@nidegen Since Carthage doesn't do dependencies for binaries, the multiple Firebase components that depend on `Protobuf.framework` need to include it.

**Comment by schlingding** [comment]
Some frameworks are fetching version 5.0.1 and some version 5.1.0.

Example:
`binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json"` is only pulling down 5.0.1

`binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json" == 5.0.0` produces
`No available version for binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json" satisfies the requirement: == 5.1.0`

The JSON file contains 5.1.0 but it seems Carthage but perhaps it does not like that is out of order.

Anyone else seeing this issue?

**Comment by nidegen** [comment]
@paulb777 i dont get it. if binaries have no dependencies (I already understood it that way) why is there an issue with a dependency (protobuf). To me it seems that carthage does check for dependencies after all, or performs some kind of processing of the binaries. i dont see why it should / a binary package is performing anything

**Comment by paulb777** [comment]
@nidegen The Carthage zip files for multiple Firebase components include the Protobuf.framework. If you're using two or more of those components, Carthage will attempt to do multiple copies to `./Carthage/Build/iOS/Protobuf.framework`.

@schlingding Does clearing the Carthage cache make a difference - `rm -rf ~/Library/Caches/org.carthage.CarthageKit`

**Comment by schlingding** [comment]
@paulb777 I have tried clearing the cache and removing the Carthage folder as well. Same results. I’m on the latest version of Carthage.

**Comment by schlingding** [comment]
@paulb777 Ran Carthage today and now it's pulling all `5.1.0` versions. Leaving the previous comment intact in case someone else runs into the issue.

**Comment by walsht** [comment]
Seeing this on "5.2.0"
```
Failed to write to /.../Carthage/Build/iOS/Protobuf.framework: Error Domain=NSCocoaErrorDomain Code=513 "“Protobuf.framework” couldn’t be removed because you don’t have permission to access it." UserInfo={NSFilePath=/.../Carthage/Build/iOS/Protobuf.framework, NSUserStringVariant=(
    Remove
), NSUnderlyingError=0x7faa64d60730 {Error Domain=NSPOSIXErrorDomain Code=66 "Directory not empty"}}
```
Only way to resolve is 
binary "https://dl.google.com/dl/firebase/ios/carthage/FirebasePerformanceBinary.json" == 5.1.0

**Comment by nidegen** [comment]
@paulb777 For the protobuf checkout bug, deleting cache with `rm -rf ~/Library/Caches/org.carthage.CarthageKit` actually fixes my issue, probably also yours @walsht 

I run it before each carthage update

**Comment by tbaranes** [comment]
Hey guys! Can you add macOS support? The `binary binary ".../firebase/ios/carthage/FirebaseAnalyticsBinary.json"` is working perfectly for iOS, but it would even greater to have a binary link for macOS as well.

**Comment by josemssilva** [comment]
Anyone had troubles updating from 5.4 to 5.6? Missing `FirebaseNanoPB` and `GoogleToolboxForMac`, and I'm unable to build the project without those.

**Comment by schlingding** [comment]
Firebase is releasing weeks faster than the updates to the binary files for carthage.

At the time of this post, I'm seeing the latest carthage binary references at 5.8.0 where 5.8.1, 5.9.0, and 5.10.0 have since been released.

Can updating these binary references be part of the Firebase iOS release process or are there further issues?

Please update soon so the community can take advantage of the bug fixes and enhancements.

Thank you!

**Comment by paulb777** [comment]
@schlingding 5.10.0 is now available via Carthage.

See #1862 to follow the status of official Carthage support

**Comment by paulb777** [comment]
Those of you interested in Firebase via Carthage may want to check out Firebase via Rome. Rome provides many of the Carthage features as well as a few additional advantages. Check out the details at https://github.com/firebase/firebase-ios-sdk/blob/master/Rome.md.  

We'd love to hear any feedback in another [issue](https://github.com/firebase/firebase-ios-sdk/issues).

**Comment by paulb777** [comment]
Closing since Carthage is now available experimentally. We'll track making it officially supported in #1862.

**Comment by zero-tolerance0** [comment]
thanks, @paulb777 clearing carthage cache solved my issue
